api/src/main/java/io/druid/data/input/impl/prefetch/RetryingInputStream.java

```diff
@@ -42,112 +51,125 @@
       T object,
       ObjectOpenFunction<T> objectOpenFunction,
       Predicate<Throwable> retryCondition,
-      int maxTry
+      int maxRetry
   ) throws IOException
   {
-    this.retryCondition = t -> {
-      Preconditions.checkNotNull(t);
+    this.object = object;
+    this.objectOpenFunction = objectOpenFunction;
+    this.retryCondition = retryCondition;
+    this.maxRetry = maxRetry;
+    this.delegate = new CountingInputStream(objectOpenFunction.open(object));
+  }
 
-      if (isConnectionReset(t)) {
+  private boolean isConnectionReset(Throwable t)
+  {
+    return (t instanceof SocketException && (t.getMessage() != null && t.getMessage().contains("Connection reset"))) ||
+           (t.getCause() != null && isConnectionReset(t.getCause()));
+  }
+
+  private void waitOrThrow(Throwable t, int i) throws IOException
+  {
+    final boolean isConnectionReset = isConnectionReset(t);
+    if (isConnectionReset || retryCondition.apply(t)) {
+      if (isConnectionReset) {
+        // Re-open the input stream on connection reset
         try {
           startOffset += delegate.getCount();
           log.info("retrying from offset[%d]", startOffset);
           delegate.close();
```

Comment:
Probably should catch exception and ignore it around this line separately

Reply:
Thanks. Fixed.